### ISSUE TITLE
Issues with PureVPN which lead to container not starting at all: Suggest a more graceful way to handle it

### ISSUE THREAD

**Issue by DarthSmeg** [haugene/docker-transmission-openvpn#1790]
## Before creating this issue I have:

**REQUIRED**
- [X] Read through the pinned issues for related problems
- [X] Searched for similar [provider issues][pi] and [container issues][ci]
- [X] [Read the documentation][rd], especially the troubleshooting section and FAQ
- [X] Tried to add as much relevant information to the issue as possible
- [X] Verified I have tried using newest release as well
- [X] Agreed that my issue will be closed if I do not follow this template and will remain closed until I complete the template

[pi]: https://github.com/haugene/vpn-configs-contrib/issues
[ci]: https://github.com/haugene/docker-transmission-openvpn/issues
[rd]: https://haugene.github.io/docker-transmission-openvpn/

### Container version & last working release

**Required, problem occurs in :** 
 
```3.3 and 3.4```

*If possible, last working version:* 
 
```latest```

### Describe the problem

**REQUIRED**
```txt
I frequently experience issues when starting the container, and have traced them to issues with the PureVPN servers. 
The symptoms is that the container does not start, and the webgui does not load. Running "docker exec -it <container-id> bash" and then "curl localhost:9091" results in a "Connection refused" error. 

Changing the OPENVPN_CONFIG= to point to a different server fixes the problem.
```
### Describe the steps you have tried to solve the problem

**REQUIRED**
```txt
I did a lot of useless troubleshooting, including deleting the container and recreating it. In the end the solution was to point the OPENVPN_CONFIG= to a different (and randomly working) server.
```

### Add your docker run command or docker-compose file or env details

**REQUIRED**
```txt
My Docker-compose file: 

 transmission-vpn:
    container_name: transmission-vpn
    image: haugene/transmission-openvpn
    cap_add:
      - NET_ADMIN
    devices:
      - /dev/net/tun
    restart: unless-stopped
    ports:
    - "9091:9091"
    dns:
      - 8.8.8.8
      - 8.8.4.4
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /mnt/store/downloads:/data
    environment:
      - OPENVPN_PROVIDER=PUREVPN
      - "OPENVPN_CONFIG=nl2-ovpn-tcp,nl2-ovpn-udp,dk2-ovpn-tcp,se2-ovpn-tcp"
      - OPENVPN_USERNAME=${VPN_USER}
      - OPENVPN_PASSWORD=${VPN_PASSWORD}
      - LOCAL_NETWORK=192.168.1.0/24
      - TZ=${TZ}
```

### Logs

 

**REQUIRED**
```txt
transmission-vpn    | Starting container with revision: de2598df732811bfb4ac25395c5f27d6fa3b0d87
transmission-vpn    | Creating TUN device /dev/net/tun
transmission-vpn    | mknod: /dev/net/tun: File exists
transmission-vpn    | Using OpenVPN provider: PUREVPN
transmission-vpn    | 3 servers found in OPENVPN_CONFIG, de2-ovpn-tcp chosen randomly
transmission-vpn    | Starting OpenVPN using config de2-ovpn-tcp.ovpn
transmission-vpn    | Setting OpenVPN credentials...
transmission-vpn    | adding route to local network 192.168.1.0/24 via 172.19.0.1 dev eth0
transmission-vpn    | Sat Mar 27 21:07:24 2021 OpenVPN 2.4.9 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Apr 20 2020
transmission-vpn    | Sat Mar 27 21:07:24 2021 library versions: OpenSSL 1.1.1g  21 Apr 2020, LZO 2.10
transmission-vpn    | Sat Mar 27 21:07:24 2021 NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
transmission-vpn    | Sat Mar 27 21:07:24 2021 Outgoing Control Channel Authentication: Using 160 bit message hash 'SHA1' for HMAC authentication
transmission-vpn    | Sat Mar 27 21:07:24 2021 Incoming Control Channel Authentication: Using 160 bit message hash 'SHA1' for HMAC authentication
transmission-vpn    | Sat Mar 27 21:07:24 2021 TCP/UDP: Preserving recently used remote address: [AF_INET]5.254.88.212:80
transmission-vpn    | Sat Mar 27 21:07:24 2021 Socket Buffers: R=[131072->131072] S=[16384->16384]
transmission-vpn    | Sat Mar 27 21:07:24 2021 Attempting to establish TCP connection with [AF_INET]5.254.88.212:80 [nonblock]
transmission-vpn    | Sat Mar 27 21:07:25 2021 TCP connection established with [AF_INET]5.254.88.212:80
transmission-vpn    | Sat Mar 27 21:07:25 2021 TCP_CLIENT link local: (not bound)
transmission-vpn    | Sat Mar 27 21:07:25 2021 TCP_CLIENT link remote: [AF_INET]5.254.88.212:80
transmission-vpn    | Sat Mar 27 21:07:32 2021 Connection reset, restarting [-1]
transmission-vpn    | Sat Mar 27 21:07:32 2021 SIGUSR1[soft,connection-reset] received, process restarting
transmission-vpn    | Sat Mar 27 21:07:32 2021 Restart pause, 3 second(s)
```

### Host system

**REQUIRED**
```txt
Running Docker on a Debian box
```

**Comment by DarthSmeg** [comment]
By the way, I tried several times to create this as a Feature request, but it was auto-closed due to template issues. I don't really know what I did wrong (I was veeery careful on round 3) :)

**Comment by DarthSmeg** [comment]
By the way, I tried several times to create this as a Feature request, but it was auto-closed due to template issues. I don't really know what I did wrong (I was veeery careful on round 3) :)

**Comment by haugene** [comment]
Hey. Sorry about the eager github bot, it's a newcomer here and we might have to tune it. And if the template seems kind of rigid it's because there are just so many issues being filed that more or less says "it doesn't work", or even better "it used to work and now it don't". So after years of asking for docker logs and run statements we are tightening up here :smile: 

Anyways... I see how this is less informative than it could be. But the question is how we should handle it. I mean, it's not us handling the connection to the server - that is OpenVPN. So the lack of "couldn't connect because the server didn't reply in a good way" is on the OpenVPN logging and not us.

Do you have any suggestions about what to do here? We have made it a bit of a point of not getting in OpenVPNs way. We do initializing of a file system, but then we relieve the shell to OpenVPN. So the container runs OpenVPN and if OpenVPN fails the container stops. We're not in the loop after OpenVPN starts.

I mean. We could change that. That we run OpenVPN and if it crashes we try to do a post-mortem and log something useful before exit'ing. But it puts us on the critical path at runtime and that's not the type of things we've done so far. We could maybe do a telnet or something to the server and port to check if something is responding and then warning if it isn't? Or just have it more explicitly mentioned in the documentation? What do you think?

This ultimately comes down to out-of-date configs and the management of that whole thing. We've recently split the configs out to a separate repository to make it easier to involve more of the community in maintaining them. I'm thinking of creating a "config checker" script that will loop through all the servers in the configs and check that something is responding. To create a "health metric" of sorts for the provider bundles.

Curious to hear your thoughts on how to take this further.

**Comment by pkishino** [comment]
Btw, you can edit your post and if it passes the checks the bot will reopen the issue.. no need to create multiple

**Comment by DarthSmeg** [comment]
Hi. I did try editing the post, several times. But in the end I gave up :)

I guess a minimum-approach here would be to add some more informative errors in the logs. The current "Connection reset, restarting [-1]" is kinda vague. In this case, I guess the server didn't respond, so a message saying why that was (no response, connection refused, etc) would help a bit. But if that is, as you say, out of your hands, then I guess this is not feasible.

A more user-friendly approach would be to spin up the web-server with a page displaying the log or some error. After all, this was working just fine for weeks until a random restart saw this error occur. I have easy access to the web-pages, but need to SSH into my server to see the logs.
